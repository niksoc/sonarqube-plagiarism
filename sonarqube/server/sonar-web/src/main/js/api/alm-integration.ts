/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import {
  getCorsJSON,
  getJSON,
  post,
  postJSON,
  requestTryAndRepeatUntil
} from 'sonar-ui-common/helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';
import { AlmLanguagesStats } from '../apps/tutorials/analyzeProject/utils';

export function bindAlmOrganization(data: { installationId: string; organization: string }) {
  return post('/api/alm_integration/bind_organization', data).catch(throwGlobalError);
}

export function getAlmAppInfo(): Promise<{ application: T.AlmApplication }> {
  return getJSON('/api/alm_integration/show_app_info').catch(throwGlobalError);
}

export interface GetAlmOrganizationResponse {
  almOrganization: T.AlmOrganization;
  boundOrganization?: T.OrganizationBase;
}

export function getAlmOrganization(data: {
  installationId: string;
}): Promise<GetAlmOrganizationResponse> {
  return requestTryAndRepeatUntil(
    () => getJSON('/api/alm_integration/show_organization', data),
    { max: 25, slowThreshold: 20 },
    () => true,
    [404]
  )
    .catch(throwGlobalError)
    .then(({ almOrganization, boundOrganization }) => ({
      almOrganization: {
        ...almOrganization,
        name: almOrganization.name || almOrganization.key
      },
      boundOrganization
    }));
}

export function getRepositories(data: {
  organization: string;
}): Promise<{ repositories: T.AlmRepository[] }> {
  return getJSON('/api/alm_integration/list_repositories', data).catch(throwGlobalError);
}

export function listUnboundApplications(): Promise<T.AlmUnboundApplication[]> {
  return getJSON('/api/alm_integration/list_unbound_applications').then(
    ({ applications }) =>
      applications.map((app: T.AlmUnboundApplication) => ({ ...app, name: app.name || app.key })),
    throwGlobalError
  );
}

export function provisionProject(data: {
  installationKeys: string[];
  organization: string;
}): Promise<{ projects: Array<{ projectKey: string }> }> {
  return postJSON('/api/alm_integration/provision_projects', {
    ...data,
    installationKeys: data.installationKeys.join(',')
  }).catch(throwGlobalError);
}

export function getGithubLanguages(url: string): Promise<AlmLanguagesStats> {
  // We don't want to throwGlobalError
  const apiUrl = url.replace('https://github.com/', 'https://api.github.com/repos/');
  return getCorsJSON(`${apiUrl}/languages`);
}
