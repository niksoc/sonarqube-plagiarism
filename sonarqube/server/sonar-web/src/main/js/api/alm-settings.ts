/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getJSON, post } from 'sonar-ui-common/helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';
import {
  AlmSettingsBindingDefinitions,
  AlmSettingsInstance,
  AzureBindingDefinition,
  AzureProjectAlmBinding,
  BitbucketBindingDefinition,
  BitbucketProjectAlmBinding,
  GithubBindingDefinition,
  GithubProjectAlmBinding,
  GitlabBindingDefinition,
  GitlabProjectAlmBinding,
  ProjectAlmBinding
} from '../types/alm-settings';

export function getAlmDefinitions(): Promise<AlmSettingsBindingDefinitions> {
  return getJSON('/api/alm_settings/list_definitions').catch(throwGlobalError);
}

export function getAlmSettings(project?: string): Promise<AlmSettingsInstance[]> {
  return getJSON('/api/alm_settings/list', { project })
    .then(({ almSettings }) => almSettings)
    .catch(throwGlobalError);
}

export function createGithubConfiguration(data: GithubBindingDefinition) {
  return post('/api/alm_settings/create_github', data).catch(throwGlobalError);
}

export function updateGithubConfiguration(data: GithubBindingDefinition & { newKey: string }) {
  return post('/api/alm_settings/update_github', data).catch(throwGlobalError);
}

export function createAzureConfiguration(data: AzureBindingDefinition) {
  return post('/api/alm_settings/create_azure', data).catch(throwGlobalError);
}

export function updateAzureConfiguration(data: AzureBindingDefinition & { newKey: string }) {
  return post('/api/alm_settings/update_azure', data).catch(throwGlobalError);
}

export function createBitbucketConfiguration(data: BitbucketBindingDefinition) {
  return post('/api/alm_settings/create_bitbucket', data).catch(throwGlobalError);
}

export function updateBitbucketConfiguration(
  data: BitbucketBindingDefinition & { newKey: string }
) {
  return post('/api/alm_settings/update_bitbucket', data).catch(throwGlobalError);
}

export function createGitlabConfiguration(data: GitlabBindingDefinition) {
  return post('/api/alm_settings/create_gitlab', data).catch(throwGlobalError);
}

export function updateGitlabConfiguration(data: GitlabBindingDefinition & { newKey: string }) {
  return post('/api/alm_settings/update_gitlab', data).catch(throwGlobalError);
}

export function deleteConfiguration(key: string) {
  return post('/api/alm_settings/delete', { key }).catch(throwGlobalError);
}

export function countBindedProjects(almSetting: string) {
  return getJSON('/api/alm_settings/count_binding', { almSetting })
    .then(({ projects }) => projects)
    .catch(throwGlobalError);
}

export function getProjectAlmBinding(project: string): Promise<ProjectAlmBinding> {
  return getJSON('/api/alm_settings/get_binding', { project });
}

export function deleteProjectAlmBinding(project: string): Promise<void> {
  return post('/api/alm_settings/delete_binding', { project }).catch(throwGlobalError);
}

export function setProjectAzureBinding(data: AzureProjectAlmBinding) {
  return post('/api/alm_settings/set_azure_binding', data).catch(throwGlobalError);
}

export function setProjectBitbucketBinding(data: BitbucketProjectAlmBinding) {
  return post('/api/alm_settings/set_bitbucket_binding', data).catch(throwGlobalError);
}

export function setProjectGithubBinding(data: GithubProjectAlmBinding) {
  return post('/api/alm_settings/set_github_binding', data).catch(throwGlobalError);
}

export function setProjectGitlabBinding(data: GitlabProjectAlmBinding) {
  return post('/api/alm_settings/set_gitlab_binding', data).catch(throwGlobalError);
}
