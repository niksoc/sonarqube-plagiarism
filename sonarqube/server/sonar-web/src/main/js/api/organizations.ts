/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getJSON, post, postJSON } from 'sonar-ui-common/helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';

export function getOrganizations(data: {
  organizations?: string;
  member?: boolean;
}): Promise<{
  organizations: T.Organization[];
  paging: T.Paging;
}> {
  return getJSON('/api/organizations/search', data).catch(throwGlobalError);
}

export function getOrganization(key: string): Promise<T.Organization | undefined> {
  return getJSON('/api/organizations/search', { organizations: key }).then(
    r => r.organizations.find((o: T.Organization) => o.key === key),
    throwGlobalError
  );
}

interface GetOrganizationNavigation {
  adminPages: T.Extension[];
  alm?: { key: string; membersSync: boolean; personal: boolean; url: string };
  canUpdateProjectsVisibilityToPrivate: boolean;
  isDefault: boolean;
  pages: T.Extension[];
}

export function getOrganizationNavigation(key: string): Promise<GetOrganizationNavigation> {
  return getJSON('/api/navigation/organization', { organization: key }).then(
    r => r.organization,
    throwGlobalError
  );
}

export function getOrganizationsThatPreventDeletion(): Promise<{
  organizations: T.Organization[];
}> {
  return getJSON('/api/organizations/prevent_user_deletion').catch(throwGlobalError);
}

export function createOrganization(
  data: T.OrganizationBase & { installationId?: string }
): Promise<T.Organization> {
  return postJSON('/api/organizations/create', data).then(r => r.organization, throwGlobalError);
}

export function updateOrganization(key: string, changes: T.OrganizationBase) {
  return post('/api/organizations/update', { key, ...changes }).catch(throwGlobalError);
}

export function deleteOrganization(key: string) {
  return post('/api/organizations/delete', { key }).catch(throwGlobalError);
}

export function searchMembers(data: {
  organization?: string;
  p?: number;
  ps?: number;
  q?: string;
  selected?: string;
}): Promise<{ paging: T.Paging; users: T.OrganizationMember[] }> {
  return getJSON('/api/organizations/search_members', data).catch(throwGlobalError);
}

export function addMember(data: {
  login: string;
  organization: string;
}): Promise<T.OrganizationMember> {
  return postJSON('/api/organizations/add_member', data).then(r => r.user, throwGlobalError);
}

export function removeMember(data: { login: string; organization: string }) {
  return post('/api/organizations/remove_member', data).catch(throwGlobalError);
}

export interface OrganizationBilling {
  nclocCount: number;
  subscription: {
    plan?: {
      maxNcloc: number;
      price: number;
    };
    nextBillingDate?: string;
    status: 'active' | 'inactive' | 'suspended';
    trial: boolean;
  };
}

export function getOrganizationBilling(organization: string): Promise<OrganizationBilling> {
  return getJSON('/api/billing/show', { organization, p: 1, ps: 1 });
}

export function setOrganizationMemberSync(data: { enabled: boolean; organization: string }) {
  return post('/api/organizations/set_members_sync', data).catch(throwGlobalError);
}

export function syncMembers(organization: string) {
  return post('/api/organizations/sync_members', { organization }).catch(throwGlobalError);
}
