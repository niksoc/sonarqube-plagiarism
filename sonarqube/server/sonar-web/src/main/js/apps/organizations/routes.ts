/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { RedirectFunction, RouterState } from 'react-router';
import { lazyLoad } from 'sonar-ui-common/components/lazyLoad';
import codingRulesRoutes from '../coding-rules/routes';
import qualityGatesRoutes from '../quality-gates/routes';
import qualityProfilesRoutes from '../quality-profiles/routes';
import webhooksRoutes from '../webhooks/routes';

const OrganizationContainer = lazyLoad(() => import('./components/OrganizationContainer'));

const routes = [
  {
    path: ':organizationKey',
    component: lazyLoad(() => import('./components/OrganizationPage')),
    childRoutes: [
      {
        indexRoute: {
          onEnter(nextState: RouterState, replace: RedirectFunction) {
            const { params } = nextState;
            replace(`/organizations/${params.organizationKey}/projects`);
          }
        }
      },
      {
        path: 'projects',
        component: OrganizationContainer,
        childRoutes: [
          { indexRoute: { component: lazyLoad(() => import('./components/OrganizationProjects')) } }
        ]
      },
      {
        path: 'issues',
        component: OrganizationContainer,
        childRoutes: [
          { indexRoute: { component: lazyLoad(() => import('../issues/components/AppContainer')) } }
        ]
      },
      {
        path: 'rules',
        component: OrganizationContainer,
        childRoutes: codingRulesRoutes
      },
      {
        path: 'members',
        component: lazyLoad(() => import('../organizationMembers/OrganizationMembersContainer'))
      },
      {
        path: 'quality_profiles',
        childRoutes: qualityProfilesRoutes
      },
      {
        path: 'quality_gates',
        component: OrganizationContainer,
        childRoutes: qualityGatesRoutes
      },
      {
        component: lazyLoad(() => import('./components/OrganizationAccessContainer')),
        childRoutes: [
          { path: 'edit', component: lazyLoad(() => import('./components/OrganizationEdit')) },
          { path: 'groups', component: lazyLoad(() => import('../groups/components/App')) },
          {
            path: 'permissions',
            component: lazyLoad(() => import('../permissions/global/components/App'))
          },
          {
            path: 'permission_templates',
            component: lazyLoad(() => import('../permission-templates/components/App'))
          },
          {
            path: 'projects_management',
            component: lazyLoad(() => import('../projectsManagement/AppContainer'))
          },
          { path: 'webhooks', childRoutes: webhooksRoutes },
          {
            path: 'extension/:pluginKey/:extensionKey',
            component: lazyLoad(() =>
              import('../../app/components/extensions/OrganizationPageExtension')
            )
          }
        ]
      }
    ]
  }
];

export default routes;
