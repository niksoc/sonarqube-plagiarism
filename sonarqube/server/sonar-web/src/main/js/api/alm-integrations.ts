/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { get, getJSON, post, postJSON } from 'sonar-ui-common/helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';
import { BitbucketProject, BitbucketRepository } from '../types/alm-integration';
import { ProjectBase } from './components';

export function setAlmPersonalAccessToken(almSetting: string, pat: string): Promise<void> {
  return post('/api/alm_integrations/set_pat', { almSetting, pat }).catch(throwGlobalError);
}

export function checkPersonalAccessTokenIsValid(almSetting: string): Promise<boolean> {
  return get('/api/alm_integrations/check_pat', { almSetting })
    .then(() => true)
    .catch(response => {
      if (response.status === 400) {
        return false;
      } else {
        return throwGlobalError(response);
      }
    });
}

export function getBitbucketServerProjects(
  almSetting: string
): Promise<{ projects: BitbucketProject[] }> {
  return getJSON('/api/alm_integrations/list_bitbucketserver_projects', { almSetting });
}

export function getBitbucketServerRepositories(
  almSetting: string,
  projectName: string
): Promise<{
  isLastPage: boolean;
  repositories: BitbucketRepository[];
}> {
  return getJSON('/api/alm_integrations/search_bitbucketserver_repos', {
    almSetting,
    projectName
  });
}

export function importBitbucketServerProject(
  almSetting: string,
  projectKey: string,
  repositorySlug: string
): Promise<{ project: ProjectBase }> {
  return postJSON('/api/alm_integrations/import_bitbucketserver_project', {
    almSetting,
    projectKey,
    repositorySlug
  }).catch(throwGlobalError);
}

export function searchForBitbucketServerRepositories(
  almSetting: string,
  repositoryName: string
): Promise<{
  isLastPage: boolean;
  repositories: BitbucketRepository[];
}> {
  return getJSON('/api/alm_integrations/search_bitbucketserver_repos', {
    almSetting,
    repositoryName
  });
}
