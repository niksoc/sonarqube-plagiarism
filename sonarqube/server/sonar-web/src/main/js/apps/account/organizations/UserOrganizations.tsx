/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { Helmet } from 'react-helmet-async';
import { connect } from 'react-redux';
import { Link } from 'react-router';
import { translate } from 'sonar-ui-common/helpers/l10n';
import {
  getAppState,
  getGlobalSettingValue,
  getMyOrganizations,
  Store
} from '../../../store/rootReducer';
import { fetchIfAnyoneCanCreateOrganizations } from './actions';
import OrganizationsList from './OrganizationsList';

interface StateProps {
  anyoneCanCreate: boolean;
  canAdmin?: boolean;
  organizations: T.Organization[];
}

interface DispatchProps {
  fetchIfAnyoneCanCreateOrganizations: () => Promise<void>;
}

interface Props extends StateProps, DispatchProps {}

interface State {
  loading: boolean;
}

class UserOrganizations extends React.PureComponent<Props, State> {
  mounted = false;
  state: State = { loading: true };

  componentDidMount() {
    this.mounted = true;
    this.props.fetchIfAnyoneCanCreateOrganizations().then(this.stopLoading, this.stopLoading);
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  stopLoading = () => {
    if (this.mounted) {
      this.setState({ loading: false });
    }
  };

  render() {
    const { anyoneCanCreate } = this.props;
    const canCreateOrganizations = !this.state.loading && (anyoneCanCreate || this.props.canAdmin);

    return (
      <div className="account-body account-container">
        <Helmet title={translate('my_account.organizations')} />

        <div className="boxed-group">
          {canCreateOrganizations && (
            <div className="clearfix">
              <div className="boxed-group-actions">
                <Link className="button" to="/create-organization">
                  {translate('create')}
                </Link>
              </div>
            </div>
          )}
          <div className="boxed-group-inner">
            {this.state.loading ? (
              <i className="spinner" />
            ) : (
              <OrganizationsList organizations={this.props.organizations} />
            )}
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state: Store): StateProps => {
  const anyoneCanCreate = getGlobalSettingValue(state, 'sonar.organizations.anyoneCanCreate');
  return {
    anyoneCanCreate: Boolean(anyoneCanCreate && anyoneCanCreate.value === 'true'),
    canAdmin: getAppState(state).canAdmin,
    organizations: getMyOrganizations(state)
  };
};

const mapDispatchToProps = {
  fetchIfAnyoneCanCreateOrganizations: fetchIfAnyoneCanCreateOrganizations as any
} as DispatchProps;

export default connect(mapStateToProps, mapDispatchToProps)(UserOrganizations);
