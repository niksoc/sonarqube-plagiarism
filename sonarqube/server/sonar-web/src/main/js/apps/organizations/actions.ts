/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { Dispatch } from 'redux';
import { translate } from 'sonar-ui-common/helpers/l10n';
import * as api from '../../api/organizations';
import { addGlobalSuccessMessage } from '../../store/globalMessages';
import * as actions from '../../store/organizations';

export function updateOrganization(key: string, changes: T.OrganizationBase) {
  return (dispatch: Dispatch<any>) => {
    return api.updateOrganization(key, changes).then(() => {
      dispatch(actions.updateOrganization(key, changes));
      dispatch(addGlobalSuccessMessage(translate('organization.updated')));
    });
  };
}

export function deleteOrganization(key: string) {
  return (dispatch: Dispatch<any>) => {
    return api.deleteOrganization(key).then(() => {
      dispatch(actions.deleteOrganization(key));
    });
  };
}
