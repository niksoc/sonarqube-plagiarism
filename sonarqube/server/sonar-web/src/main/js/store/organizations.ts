/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { omit, uniq, without } from 'lodash';
import { combineReducers } from 'redux';
import { ActionType } from './utils/actions';

type ReceiveOrganizationsAction =
  | ActionType<typeof receiveOrganizations, 'RECEIVE_ORGANIZATIONS'>
  | ActionType<typeof receiveMyOrganizations, 'RECEIVE_MY_ORGANIZATIONS'>;

type Action =
  | ReceiveOrganizationsAction
  | ActionType<typeof createOrganization, 'CREATE_ORGANIZATION'>
  | ActionType<typeof updateOrganization, 'UPDATE_ORGANIZATION'>
  | ActionType<typeof deleteOrganization, 'DELETE_ORGANIZATION'>;

export interface State {
  byKey: T.Dict<T.Organization>;
  my: string[];
}

export function receiveOrganizations(organizations: T.Organization[]) {
  return { type: 'RECEIVE_ORGANIZATIONS', organizations };
}

export function receiveMyOrganizations(organizations: T.Organization[]) {
  return { type: 'RECEIVE_MY_ORGANIZATIONS', organizations };
}

export function createOrganization(organization: T.Organization) {
  return { type: 'CREATE_ORGANIZATION', organization };
}

export function updateOrganization(key: string, changes: T.OrganizationBase) {
  return { type: 'UPDATE_ORGANIZATION', key, changes };
}

export function deleteOrganization(key: string) {
  return { type: 'DELETE_ORGANIZATION', key };
}

function onReceiveOrganizations(state: State['byKey'], action: ReceiveOrganizationsAction) {
  const nextState = { ...state };
  action.organizations.forEach(organization => {
    nextState[organization.key] = { ...state[organization.key], ...organization };
  });
  return nextState;
}

function byKey(state: State['byKey'] = {}, action: Action): State['byKey'] {
  switch (action.type) {
    case 'RECEIVE_ORGANIZATIONS':
    case 'RECEIVE_MY_ORGANIZATIONS':
      return onReceiveOrganizations(state, action);
    case 'CREATE_ORGANIZATION':
      return {
        ...state,
        [action.organization.key]: {
          ...action.organization,
          actions: { ...(action.organization.actions || {}), admin: true }
        }
      };
    case 'UPDATE_ORGANIZATION':
      return {
        ...state,
        [action.key]: {
          ...state[action.key],
          key: action.key,
          ...action.changes
        }
      };
    case 'DELETE_ORGANIZATION':
      return omit(state, action.key);
    default:
      return state;
  }
}

function my(state: State['my'] = [], action: Action): State['my'] {
  switch (action.type) {
    case 'RECEIVE_MY_ORGANIZATIONS':
      return uniq([...state, ...action.organizations.map(o => o.key)]);
    case 'CREATE_ORGANIZATION':
      return uniq([...state, action.organization.key]);
    case 'DELETE_ORGANIZATION':
      return without(state, action.key);
    default:
      return state;
  }
}

export default combineReducers({ byKey, my });

export function getOrganizationByKey(state: State, key: string) {
  return state.byKey[key];
}

export function getMyOrganizations(state: State) {
  return state.my.map(key => getOrganizationByKey(state, key));
}

export function areThereCustomOrganizations(state: State) {
  return Object.keys(state.byKey).length > 1;
}
