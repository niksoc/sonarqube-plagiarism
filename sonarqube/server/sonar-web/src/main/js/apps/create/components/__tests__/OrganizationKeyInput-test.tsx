/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { waitAndUpdate } from 'sonar-ui-common/helpers/testUtils';
import { getOrganization } from '../../../../api/organizations';
import OrganizationKeyInput from '../OrganizationKeyInput';

jest.mock('../../../../api/organizations', () => ({
  getOrganization: jest.fn().mockResolvedValue(undefined)
}));

beforeEach(() => {
  (getOrganization as jest.Mock<any>).mockClear();
});

it('should render correctly', async () => {
  const wrapper = shallow(<OrganizationKeyInput initialValue="key" onChange={jest.fn()} />);
  expect(wrapper).toMatchSnapshot();
  wrapper.setState({ touched: true });
  await waitAndUpdate(wrapper);
  expect(wrapper.find('ValidationInput').prop('isValid')).toBe(true);
});

it('should not display any status when the key is not defined', async () => {
  const wrapper = shallow(<OrganizationKeyInput onChange={jest.fn()} />);
  await waitAndUpdate(wrapper);
  expect(wrapper.find('ValidationInput').prop('isInvalid')).toBe(false);
  expect(wrapper.find('ValidationInput').prop('isValid')).toBe(false);
});

it('should have an error when the key is invalid', async () => {
  const wrapper = shallow(
    <OrganizationKeyInput initialValue="KEy-with#speci@l_char" onChange={jest.fn()} />
  );
  await waitAndUpdate(wrapper);
  expect(wrapper.find('ValidationInput').prop('isInvalid')).toBe(true);
});

it('should have an error when the key already exists', async () => {
  (getOrganization as jest.Mock<any>).mockResolvedValue({});
  const wrapper = shallow(<OrganizationKeyInput initialValue="" onChange={jest.fn()} />);
  await waitAndUpdate(wrapper);
  expect(wrapper.find('ValidationInput').prop('isInvalid')).toBe(true);
});
