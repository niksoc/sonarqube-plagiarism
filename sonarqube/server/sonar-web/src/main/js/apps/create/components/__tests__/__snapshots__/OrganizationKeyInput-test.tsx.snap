// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<ValidationInput
  id="organization-key"
  isInvalid={false}
  isValid={false}
  label="onboarding.create_organization.organization_name"
  required={true}
>
  <div
    className="display-inline-flex-baseline"
  >
    <span
      className="little-spacer-right"
    >
      localhost/organizations/
    </span>
    <input
      autoFocus={true}
      className="input-super-large"
      id="organization-key"
      maxLength={255}
      onChange={[Function]}
      type="text"
      value="key"
    />
  </div>
</ValidationInput>
`;
